src/model/calculations.ts

```diff
@@ -100,11 +101,154 @@ export class OptionsRemainingCalculator {
   }
 }
 
+interface StockClass {
+  object_type: string;
+  id: string;
+  name: string;
+  class_type: string;
+  default_id_prefix?: string;
+  current_shares_authorized?: string;
+  board_approval_date?: string;
+  votes_per_share?: number;
+  conversion_rights?: StockClassConversionRight[];
+}
+
+interface StockClassConversionRight {
+  conversion_mechanism: {
+    type: string;
+    ratio: {
+      numerator: string;
+      denominator: string;
+    };
+  };
+  rounding_type?: string;
+  converts_to_stock_class_id?: string;
+}
+
+export class ConversionRatioCalculator {
+  private stockClasses: StockClass[] = [];
+
+  public findRatio(preferredStockClassId: string): {
+    ratio: number;
+    lowestVotesPerShare: number | undefined;
+    path: StockClass[];
+  } {
+    const visited: Set<string> = new Set();
+    let lowestVotesPerShare: number | undefined;
+    let path: StockClass[] = [];
+
+    // Find the root Preferred Stock Class
+    const rootPreferredStockClass = this.stockClasses.find(
+      (cls) =>
+        cls.id === preferredStockClassId && cls.class_type === "PREFERRED"
+    );
+    if (rootPreferredStockClass === undefined) {
+      throw new Error("Preferred Stock Class not found.");
+    }
+
+    const queue: { stockClass: StockClass; currentPath: StockClass[] }[] = [
+      { stockClass: rootPreferredStockClass, currentPath: [] },
+    ];
+
+    // multiple conversion rights
+    while (queue.length > 0) {
+      const dequeuedClass = queue.shift();
+      if (dequeuedClass) {
+        const { stockClass, currentPath } = dequeuedClass;
+        if (!visited.has(stockClass.id)) {
+          visited.add(stockClass.id);
```

Comment:
**Non-blocking:** At a glance, this while loop looks more complex than necessary, but that might just be because it's long and could use some well-named functions for clarity. Will be curious to see how this evolves as we add more cases to the tests.